File: frontend/src/services/geminiService.js
export async function askGemini(prompt, history = []) {
  const key = import.meta.env.VITE_GEMINI_KEY
  if (!key) throw new Error('VITE_GEMINI_KEY is missing in frontend/.env — restart Vite after adding it.')

  const contents = [
    ...history,
    { role: 'user', parts: [{ text: prompt }] },
  ]

  const models = ["gemini-flash-latest", "gemini-2.5-flash", "gemini-3.5-flash"]
  const controller = new AbortController()
  const timer = setTimeout(() => controller.abort(), 30000)

  try {
    for (const model of models) {
      const url = `https://generativelanguage.googleapis.com/v1beta/models/${model}:generateContent?key=${key}`
      const response = await fetch(url, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        signal: controller.signal,
        body: JSON.stringify({
          contents,
          generationConfig: { temperature: 0.7, maxOutputTokens: 1000 },
        }),
      })
      const data = await response.json()
      if (response.ok && data.candidates?.[0]?.content?.parts?.[0]?.text) {
        return data.candidates[0].content.parts[0].text
      }
      if (data.error || !response.ok) {
        console.warn(`Gemini ${model} failed:`, data.error?.message || response.statusText);
        continue;
      }
    }
    
    // Fallback to Groq if all Gemini models fail or face high demand
    const { askGroq } = await import('./groqService.js');
    return await askGroq(prompt, '', history);
    
  } catch (err) {
    if (err.name === 'AbortError') throw new Error('AI timed out (30s). Please try again.')
    throw err
  } finally {
    clearTimeout(timer)
  }
}

File: backend/services/gemini_service.py
import os
import httpx

def ask_ai(prompt: str) -> str:
    gemini_key = os.getenv("GEMINI_API_KEY", "")
    if not gemini_key:
        return "AI is temporarily unavailable."
    
    models = ["gemini-flash-latest", "gemini-2.5-flash", "gemini-3.5-flash"]
    for model in models:
        url = f"https://generativelanguage.googleapis.com/v1beta/models/{model}:generateContent?key={gemini_key}"
        res = httpx.post(url, json={"contents": [{"parts": [{"text": prompt}]}]}, timeout=30)
        data = res.json()
        if res.status_code == 200:
            return data["candidates"][0]["content"]["parts"][0]["text"]
        if data.get("error", {}).get("code") != 404:
            return "AI error: " + data.get("error", {}).get("message", "Unknown error")
    return "AI error: No available Gemini model endpoint found."